ik.py

```diff
@@ -0,0 +1,265 @@
+
+    torso_pos, torso_orn = p.getBasePositionAndOrientation(robot_id)
+    inv_torso_pos, inv_torso_orn = p.invertTransform(torso_pos, torso_orn)
+    target_pos_local = p.multiplyTransforms(inv_torso_pos, inv_torso_orn, goal_pos, [0, 0, 0, 1])[0]
+
+    movable_joints = [
+        j for j in range(p.getNumJoints(robot_id)) if p.getJointInfo(robot_id, j)[2] != p.JOINT_FIXED
+    ]
+
+    current_positions = [p.getJointState(robot_id, j)[0] for j in movable_joints]
+    solution = p.calculateInverseKinematics(
+        robot_id,
+        ee_id,
+        target_pos_local,
+        currentPositions=current_positions,
+        lowerLimits=lower_limits,
+        upperLimits=upper_limits,
+        jointRanges=joint_ranges,
+    )
+
+    # Safety check: Ensure the solution is within acceptable limits
+    is_solution_safe = all(
+        abs(current_positions[i] - solution[i]) <= MAX_JOINT_DELTA
+        for i in range(len(solution))
+    )
+
+    if not is_solution_safe:
+        print("IK solution exceeds safety limits, skipping update.")
+        return np.array(current_positions)
+
+    for i, val in enumerate(solution):
+        joint_name = list(initial_positions.keys())[i]
+        if joint_name in ee_chain:
+            p.resetJointState(robot_id, joint_info[joint_name]["index"], val)
+
+    actual_pos, _ = p.getLinkState(robot_id, ee_id)[:2]
+    error = np.linalg.norm(np.array(goal_pos) - np.array(actual_pos))
+
+    print(f"Arm: {arm}, Goal : {error}")
+
+    return np.array(solution)
+
+def read_controller_input() -> tuple[float, float, float, bool]:
+    """Reads input from the Logitech controller and returns changes in position and button state."""
+    x_change, y_change, z_change = 0, 0, 0
+    toggle_button_pressed = False
+    events = get_gamepad()
+    for event in events:
+        if event.ev_type == 'Absolute':
+            if event.code == 'ABS_X':  # Left stick horizontal
+                x_change = event.state / 32768.0  # Normalize the value
+            elif event.code == 'ABS_Y':  # Left stick vertical
+                y_change = event.state / 32768.0
+            elif event.code == 'ABS_RZ':  # Right stick vertical
+                z_change = event.state / 32768.0
+        elif event.ev_type == 'Key' and event.code == 'BTN_SOUTH' and event.state == 1:
+            toggle_button_pressed = True
+    return x_change, y_change, z_change, toggle_button_pressed
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--real", action="store_true")
+    args = parser.parse_args()
+
+    if args.real:
+        left_arm, right_arm = initialize_robot()
+
+    # Initialize PyBullet
+    p.connect(p.GUI)
+    p.setAdditionalSearchPath(pybullet_data.getDataPath())
+    robot_id = p.loadURDF("urdf/stompy_pro/robot.urdf", [0, 0, 0], useFixedBase=True)
+    p.setGravity(0, 0, -9.81)
+
+    goal_pos_left = np.array([-0.25, -0.25, 0])
```

Comment:
This could be first [0, 0, 0] and the arms being loose and looking downwards. Then we don't have to worry about the offset much and just make sure signs are correct.